File: ImperatorToCK3/Source/Mappers/DeathReasonMapper/DeathReasonMapper.cpp
#include "DeathReasonMapper.h"
#include "DeathReasonMapping.h"
#include "Log.h"
#include "ParserHelpers.h"
#include "CommonRegexes.h"



mappers::DeathReasonMapper::DeathReasonMapper() {
	Log(LogLevel::Info) << "-> Parsing death reason mappings.";
	registerKeys();
	parseFile("configurables/deathMappings.txt");
	clearRegisteredKeywords();
	Log(LogLevel::Info) << "<> Loaded " << impToCK3ReasonMap.size() << " death reason links.";
}


mappers::DeathReasonMapper::DeathReasonMapper(std::istream& theStream) {
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}


void mappers::DeathReasonMapper::registerKeys() {
	registerKeyword("link", [this](std::istream& theStream) {
		const DeathReasonMapping mapping(theStream);
		for (const auto& impReason: mapping.impReasons) {
			if (mapping.ck3Reason)
				impToCK3ReasonMap.emplace(impReason, *mapping.ck3Reason);
		}
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}


std::optional<std::string> mappers::DeathReasonMapper::getCK3ReasonForImperatorReason(const std::string& impReason) const {
	const auto& mapping = impToCK3ReasonMap.find(impReason);
	if (mapping != impToCK3ReasonMap.end())
		return mapping->second;
	return std::nullopt;
}

File: ImperatorToCK3/Mappers/DeathReason/DeathReasonMapper.cs
using commonItems;
using System.Collections.Generic;

namespace ImperatorToCK3.Mappers.DeathReason;

public sealed class DeathReasonMapper {
	public DeathReasonMapper() {
		Logger.Info("Parsing death reason mappings...");
		var parser = new Parser();
		RegisterKeys(parser);
		parser.ParseFile("configurables/deathMappings.txt");
		Logger.Info($"Loaded {irToCK3ReasonMap.Count} death reason links.");

		Logger.IncrementProgress();
	}
	public DeathReasonMapper(BufferedReader reader) {
		var parser = new Parser();
		RegisterKeys(parser);
		parser.ParseStream(reader);
	}
	public string? GetCK3ReasonForImperatorReason(string irReason) {
		return irToCK3ReasonMap.TryGetValue(irReason, out var value) ? value : null;
	}

	private void RegisterKeys(Parser parser) {
		parser.RegisterKeyword("link", reader => {
			var mapping = new DeathReasonMapping(reader);
			if (mapping.Ck3Reason is null) {
				return;
			}

			foreach (var impReason in mapping.ImperatorReasons) {
				irToCK3ReasonMap.Add(impReason, mapping.Ck3Reason);
			}
		});
	}
	private readonly Dictionary<string, string> irToCK3ReasonMap = new();
}